### ISSUE TITLE
Coordinate method fails if `SkyCoord` contains `MaskedQuantity` fields

### ISSUE THREAD

**Issue by hamogu** [astropy/astropy#13041]
### Description

`apply_space_motion` on a coordinate fails, if that Coordinate contains `astropy.utils.masked.core.MaskedQuantity` fields. If, instead, the coordinate is only made with normal quantities everything works.
I admit, that the setup in the minimal example below looks a little obscure. That's only because I'm trying to give an example here that does not rely on other packages.  In practice, I'm getting the coordinates and proper motions from an astroquery call to Simbad. That results in a masked table. Then, I convert that result to a `QTable` because I need to do some unit-aware arithmetic on columns that are not directly related to this problem and so I end up with `MaskedQuantity` after conversion to the `QTable`. 

This particular problem in new in 5.0 because `MAskedColumns` were not converted  automatically to `MaskedQuantities` before.

### Expected behavior

`apply_space_motion` should work for any valid coordinate object, independent on how exactly that object was initialized. Note that in the minimal example below, I have two coordinates. In the first row, all numbers are provided, while in the second row some are masked. Yet, even a coordinate transform on the first row only fails. I would not be surprised to see the second row fail (although I personally would prefer a to assume that a masked quantity is treated as a no-op because that way I can run `apply_space_motion` on my entire catalog even if not all objects have known proper motions), but at the very least the first one should work.

### Actual behavior

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Input In [115], in <cell line: 1>()
----> 1 example[0]['coord'].apply_space_motion(new_obstime=Time('2020-10-22T12:23:45'))

File /nfs/melkor/d1/guenther/soft/mambaforge/envs/ciao-4.14/lib/python3.9/site-packages/astropy/coordinates/sky_coordinate.py:791, in SkyCoord.apply_space_motion(self, new_obstime, dt)
    788 except u.UnitConversionError:  # No RV
    789     rv = 0.
--> 791 starpm = erfa.pmsafe(icrsrep.lon.radian, icrsrep.lat.radian,
    792                      icrsvel.d_lon.to_value(u.radian/u.yr),
    793                      icrsvel.d_lat.to_value(u.radian/u.yr),
    794                      plx, rv, t1.jd1, t1.jd2, t2.jd1, t2.jd2)
    796 if parallax_zero:
    797     new_distance = None

File /nfs/melkor/d1/guenther/soft/mambaforge/envs/ciao-4.14/lib/python3.9/site-packages/erfa/core.py:4991, in pmsafe(ra1, dec1, pmr1, pmd1, px1, rv1, ep1a, ep1b, ep2a, ep2b)
   4855 def pmsafe(ra1, dec1, pmr1, pmd1, px1, rv1, ep1a, ep1b, ep2a, ep2b):
   4856     """
   4857     Star proper motion:  update star catalog data for space motion, with
   4858     special handling to handle the zero parallax case.
   (...)
   4989 
   4990     """
-> 4991     ra2, dec2, pmr2, pmd2, px2, rv2, c_retval = ufunc.pmsafe(
   4992         ra1, dec1, pmr1, pmd1, px1, rv1, ep1a, ep1b, ep2a, ep2b)
   4993     check_errwarn(c_retval, 'pmsafe')
   4994     return ra2, dec2, pmr2, pmd2, px2, rv2

File /nfs/melkor/d1/guenther/soft/mambaforge/envs/ciao-4.14/lib/python3.9/site-packages/astropy/utils/masked/core.py:743, in MaskedNDArray.__array_ufunc__(self, ufunc, method, *inputs, **kwargs)
    739         mask = result_masks if len(result_masks) > 1 else result_masks[0]
    741 elif method == '__call__':
    742     # Regular ufunc call.
--> 743     mask = self._combine_masks(masks, out=out_mask)
    745 elif method == 'outer':
    746     # Must have two arguments; adjust masks as will be done for data.
    747     assert len(masks) == 2

File /nfs/melkor/d1/guenther/soft/mambaforge/envs/ciao-4.14/lib/python3.9/site-packages/astropy/utils/masked/core.py:661, in MaskedNDArray._combine_masks(self, masks, out)
    659 out = np.logical_or(masks[0], masks[1], out=out)
    660 for mask in masks[2:]:
--> 661     np.logical_or(out, mask, out=out)
    662 return out

TypeError: return arrays must be of ArrayType


```

### Steps to Reproduce

Note that the error does not appear, if I set `masked=False` in the creation of the table. In that case, the proper motion columns are converted `astropy.units.quantity.Quantity` in the `QTable` and masked fields are simply filled with nan.

```python
import numpy as np
from astropy.table import Table, QTable
from astropy.coordinates import SkyCoord
from astropy.time import Time
import astropy.units as u
example = Table({'ra': [1, 2] *u.deg, 'dec':[3,4] * u.deg,
                 'pm_ra_cosdec': np.ma.array([5, np.ma.masked])*u.mas/u.yr, 
                 'pm_dec': np.ma.array([1, np.ma.masked])*u.mas/u.yr,
                 'obstime': Time(['2000-10-22T12:23:45', '2000-10-22T12:23:45'])}, masked=True)
example = QTable(example)
example['coord'] = SkyCoord(**example)
example[0]['coord'].apply_space_motion(new_obstime=Time('2020-10-22T12:23:45'))
```

### System Details

Linux-3.10.0-1062.12.1.el7.x86_64-x86_64-with-glibc2.17
Python 3.9.12 | packaged by conda-forge | (main, Mar 24 2022, 23:25:59) 
[GCC 10.3.0]
Numpy 1.22.3
pyerfa 2.0.0.1
astropy 5.0.3
Scipy 1.8.0
Matplotlib 3.5.1

**Comment by nstarman** [comment]
Ping @mhvk

**Comment by mhvk** [comment]
@hamogu - that does look like a bug in `Masked` - do I see correctly that this was for a scalar. Does it also fail if the object is an array? (Possibly of size 1).

**Comment by hamogu** [comment]
Good question. If I try to run it on an array of 2 (i.e. both rows of the table in my minimal example), I get a different error:
```
example['coord'].apply_space_motion(new_obstime=Time('2020-10-22T12:23:45'))

/Users/guenther/mambaforge/envs/i386ciao-4.14/lib/python3.9/site-packages/astropy/utils/masked/core.py:778: RuntimeWarning: invalid value encountered in pmsafe
  result = getattr(ufunc, method)(*unmasked, **kwargs)

---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
/var/folders/r7/f0qh27rn207bwwvt3s7v5gh40000gn/T/ipykernel_17366/3573469047.py in <module>
     10 example = QTable(example)
     11 example['coord'] = SkyCoord(**example)
---> 12 example['coord'].apply_space_motion(new_obstime=Time('2020-10-22T12:23:45'))

~/mambaforge/envs/i386ciao-4.14/lib/python3.9/site-packages/astropy/coordinates/sky_coordinate.py in apply_space_motion(self, new_obstime, dt)
    789             rv = 0.
    790 
--> 791         starpm = erfa.pmsafe(icrsrep.lon.radian, icrsrep.lat.radian,
    792                              icrsvel.d_lon.to_value(u.radian/u.yr),
    793                              icrsvel.d_lat.to_value(u.radian/u.yr),

~/mambaforge/envs/i386ciao-4.14/lib/python3.9/site-packages/erfa/core.py in pmsafe(ra1, dec1, pmr1, pmd1, px1, rv1, ep1a, ep1b, ep2a, ep2b)
   4991     ra2, dec2, pmr2, pmd2, px2, rv2, c_retval = ufunc.pmsafe(
   4992         ra1, dec1, pmr1, pmd1, px1, rv1, ep1a, ep1b, ep2a, ep2b)
-> 4993     check_errwarn(c_retval, 'pmsafe')
   4994     return ra2, dec2, pmr2, pmd2, px2, rv2
   4995 

~/mambaforge/envs/i386ciao-4.14/lib/python3.9/site-packages/erfa/core.py in check_errwarn(statcodes, func_name)
    138             statcodes = statcodes[statcodes > 0]
    139 
--> 140         warncodes = numpy.unique(statcodes)
    141 
    142         warncounts = dict([(w, numpy.sum(statcodes == w)) for w in warncodes])

<__array_function__ internals> in unique(*args, **kwargs)

TypeError: no implementation found for 'numpy.unique' on types that implement __array_function__: [<class 'astropy.utils.masked.core.MaskedNDArray'>]
```

**Comment by mhvk** [comment]
Ha!  OK, so now there are two bugs, one assuming we can write to the output (which is true only for an array) and another that we don't support `np.unique`... Thankfully, I actually have a branch that implements the latter -- will just have to rebase it and see why I didn't push it yet...

Sorry that this change to `Table` to be able to hold masked quantities is breaking so much. Clearly, we didn't quite think through how much *in*direct effect this would have. And here I was being so happy with how relatively easy the change was... Of course, masking never worked for `SkyCoord` before except before we just didn't get to notice that since we never fed masked values into `SkyCoord`...

**Comment by hamogu** [comment]
> Of course, masking never worked for SkyCoord before except before we just didn't get to notice that since we never fed masked values into SkyCoord.

And I think that's exactly how I will work around this: Fill with Nan so I get unmasked quantities before I turn that into a SkyCoord - but I agree that it's not the seamless experience between tables, quantities and coordinates that I had been hoping for. It also tells me that many people still use astropy in siloed pieces  - quantities for some problem, tables for another, with no communication between or just with plain numyy arrays to move data form one part of the analysis to the next. Otherwise, we'd have more reports of this type...

**Comment by mhvk** [comment]
@hamogu - I think you are right that most people still use only parts of astropy - you're on the front and suffering for it. But at least this particular problem should be fairly easy to fix!

**Comment by hamogu** [comment]
> Of course, masking never worked for SkyCoord before except before we just didn't get to notice that since we never fed masked values into SkyCoord...

Yes - but in my real use case, I have a maskedQuantity where not a single row is masked in the columns that I care for here (i.e. they all have all the numbers I need to do a coordinate, but they are missing e.g. a B_mag. Because of that missing B_mag the entire table is converted to a maskedQuantity. So, I really want to use a SkyCoord where all fields are valid number - and that case *did* work before.

**Comment by mhvk** [comment]
Rather belatedly, a fix for this issue is in #16224.
